### ISSUE TITLE
Thoughts about performance

### ISSUE THREAD

**Issue by eads** [newsdev/nyt-ap-elections#113]
All the performance #s here are on Eads’ MacBook with a fast Fios connection using the 2012-11-06 election data. 

tldr;

In order of magnitude of performance improvement:

* Refactor or eliminate set_fields
* Don't parse dates (at least in db-ready output)
* Use a faster json decoder
* Pursue additional incremental improvements, like faster bulk loading and parallel processing loops

**Eliminate set_fields**

All the string processing is slooow. Eliminating cast to unicode saves 10+ seconds, eliminating strip saves 5+, eliminating lower saves another 5+. 

Best option: explicitly set fields from kwargs in constructor. 

Second best option: pare down set_fields to a `lower()` call and `setattr`. 

**Date parsing**

Current date parser is using dateutil parse which is slow. But since we serialize back out to unparsed dates for the db, I don't see any reason to make date parsing a mandatory parsing step. It might be nice to make it optional. But Postgres COPY is so much faster than python at handling the dates and it only needs to happen once. This should save 10+ seconds. 

**Json parsing**

Using ujson instead of standard Python json library shaves about 5 seconds off. It's not a huge improvement but it is trivial to implement. 

**Parallelize processing loops**

This is just a hunch, but I suspect the parallelizeing the main loops that create the results list et al will lead to nice performance improvements that scale with the number of processors. 

The same might apply to the loops which write to stdout (and since order doesn't matter, that's fine) but I worry about logjams with threads waiting to write and portability between different OSes and versions.

All in all, I'm guessing parallelizing loops will shave a few seconds off the time. 

**Implementation**

We need to experiment with writing csv to disk and loading with the Postgres bulk loader vs using COPY. Right now it takes about 5s to load all results. The pg bulk loader is faster and supports a multiprocess mode with reader and writer processes. I imagine we could get loading down to less than 2s this way. 

**Network**

Sadly the AP takes 6-8s to return data. There's nothing we can do about this. I’d like to dig in and see when the data starts coming and how fast their servers can transfer the data, but it’s worth noting this time appears to be fairly constant on my fast home connection and my less fast work connection. These are bottlenecks we can only hope the AP will improve.

**Conclusion**

Variations of the first three tactics led to an improvement in parsing time from 55-60s to 18-22s on my system. I don't think getting sub 15s from api call to db is realistic, but comfortably under 30s is entirely feasible. If AP can speed up the API side, we probably could get under 15s. And we could experiment with partial results calls.

**Comment by jeremyjbowers** [comment]
This is wonderful. I don't immediately see anything that sets off my "not worth it" alarm. Will also ask my people if we need national-only results, which could also minimize the size of the JSON we're downloading.
